Add ETCD debug info (#555)
SUSE/doc-caasp: adoc/admin-troubleshooting-etcd.adoc

= ETCD Troubleshooting

== Introduction

This document aims to describe debugging a {etcd} cluster.

> The required etcd logs are part of the `supportconfig`, a utility that collects are the required information for debugging a problem. The rest of the document provides information on how you can obtain these information manually.

Reviewer: ```suggestion
The required etcd logs are part of the `supportconfig`, a utility that collects all the required information for debugging a problem. The rest of the document provides information on how you can obtain these information manually.
```

Author: Done

== ETCD container

ETCD is a distributed reliable key-value store for the most critical data of a distributed system. It is running **only on the master** nodes in a form a container application. For instance, in a cluster with 3 master nodes, it is expected
to have 3 etcd instances as well:

[source,bash]
----

> kubectl get pods -n kube-system -l component=etcdNAME

Reviewer: Ouch, missing break line here between `etcd` and `NAME`

Author: Thanks for spotting this!

etcd-vm073033.qa.prv.suse.net   1/1     Running   1          7d
----

The specific configuration which `etcd` is using to start, is the following:

[source,bash]
----
etcd \
      --advertise-client-urls=https://<YourMasterNodeIP>:2379 \
      --cert-file=/etc/kubernetes/pki/etcd/server.crt  \
      --client-cert-auth=true --data-dir=/var/lib/etcd \
      --initial-advertise-peer-urls=https://<YourMasterNodeIP>:2380 \
      --initial-cluster=vm072050.qa.prv.suse.net=https://<YourMasterNodeIP>:2380 \
      --key-file=/etc/kubernetes/pki/etcd/server.key \
      --listen-client-urls=https://127.0.0.1:2379,https://<YourMasterNodeIP>:2379 \
      --listen-peer-urls=https://<YourMasterNodeIP>:2380 \
      --name=vm072050.qa.prv.suse.net \
      --peer-cert-file=/etc/kubernetes/pki/etcd/peer.crt \
      --peer-client-cert-auth=true \
      --peer-key-file=/etc/kubernetes/pki/etcd/peer.key \
      --peer-trusted-ca-file=/etc/kubernetes/pki/etcd/ca.crt \
      --snapshot-count=10000 --trusted-ca-file=/etc/kubernetes/pki/etcd/ca.crt
----

[NOTE]
====
For more information related to ETCD, we **highly** recommend you to read https://etcd.io/docs/v3.4.0/faq/[ETCD FAQ] page.
====

== logging

> Since `etcd` is running in a container, that means it is no more controlled by `systemd`, thus any commands related to that (e.g. `journalctl`) will fail, therefore you need to use container debugging approach instead.

Reviewer: ```suggestion
Since `etcd` is running in a container, that means it is not controlled by `systemd`, thus any commands related to that (e.g. `journalctl`) will fail, therefore you need to use container debugging approach instead.
```

Author: Thanks

[NOTE]
====

> To use the following commands, you need to connected (e.g. via SSH) to the master node where the etcd pods are running.

Reviewer: ```suggestion
To use the following commands, you need to connect (e.g. via SSH) to the master nodes where the etcd pods are running.
```

or

```suggestion
To use the following commands, you need to connect (e.g. via SSH) to the master node where the etcd pod is running.
```

Author: Done

====

To see the `etcd` logs, connect to a {kube} master node and then run as root:
[source,bash]
----
ssh sles@<master-node>
sudo bash # connect as root
etcdcontainer=$(crictl ps --label io.kubernetes.container.name=etcd --quiet)
crictl logs -f $etcdcontainer
----

== etcdctl

> `etcdctl` is a command line client for `etcd`. The new version of CaaSP is using the `v3` API. For that, you need to make sure to set environment variable `ETCDCTL_API=3` beforing using it. Apart from that, you need to provide the required keys and certificates for the authentication and authorization, via `ETCDCTL_CACERT`, `ETCDCTL_CERT` and `ETCDCTL_KEY`. Last but not least, you need to also specify the endpoint via `ETCDCTL_ENDPOINTS`.

Reviewer: ```suggestion
`etcdctl` is a command line client for `etcd`. The new version of CaaSP is using the `v3` API. For that, you need to make sure to set environment variable `ETCDCTL_API=3` before using it. Apart from that, you need to provide the required keys and certificates for authentication and authorization, via `ETCDCTL_CACERT`, `ETCDCTL_CERT` and `ETCDCTL_KEY` environment variables. Last but not least, you need to also specify the endpoint via `ETCDCTL_ENDPOINTS` environment variable.
```

Author: Done
